crypto/x509/v3_purp.c

```diff
@@ -117,6 +120,16 @@ int X509_PURPOSE_get_count(void)
     return sk_X509_PURPOSE_num(xptable) + X509_PURPOSE_COUNT;
 }
 
+/* find smallest identifier not yet taken - note there might be gaps */
+int X509_PURPOSE_get_unused_id(ossl_unused OSSL_LIB_CTX *libctx)
+{
+    int id = X509_PURPOSE_MAX + 1;
+
+    while (X509_PURPOSE_get_by_id(id) != -1)
+        id++;
+    return id; /* is guaranteed to be unique and > X509_PURPOSE_MAX and != 0 */
+}
+
```

Comment:
Can using this function cause a thread  issue? You want to register purpose ids dynamically, right?

Reply:
Yes, `X509_PURPOSE_add()` and (therefore also its new helper function `X509_PURPOSE_get_unused_id()`) is about adding new purposes to a static structure potentially shared among threads,
and strictly speaking this can suffer from race conditions.

This inherent problem with `X509_PURPOSE_add()` has been pre-existing for the X509_PURPOSE mechanism, and this PR is not aiming to solve it.
Looks like nobody has been caring about this problem because hardly anyone uses this API.
Therefore I believe it is very unlikely that an actual thread issue will occur.

Reply:
May I ask you to document this issue (in a separate PR if you think it's better)?

Reply:
Since this PR anyway adds respective doc to  `X509_check_purpose.pod`,
I've added the following section there:
```
=head1 BUGS

The X509_PURPOSE implementation so far is not thread-safe.
There may be race conditions retrieving purpose information while
X509_PURPOSE_add() or X509_PURPOSE_cleanup() is being called.
```